### ISSUE TITLE
Consul Back-end Error Releasing Lock

### ISSUE THREAD

**Issue by practicalint** [hashicorp/terraform#14312]
I have been evaluating and configuring our terraform implementation with AWS resources. I have had success working with local and S3 backends. However, I have been struggling with using a consul backend, which I think would be Hashicorp’s preference and we think is preferred as far as keeping CSP agnostic.
 
I get the backend defined, initialized, a token created and used for accessing it (some difficulties getting there, but seems good now). But I get this error when creating an AWS instance. It occurs if the provisioning takes a while, like gets help up connecting through a bastion server and also occurs at the end of a successful completion. The state from the consul kv looks good for the instance with an tainted = false value. Terraform show looks fine. Running terraform .093 (as .094 has an issue with tags that affects us). I don’t find much when searching for similar issues. 

### Affected Resource(s) 
Any execution with Consul back-end in place, however, terraform runs appear to work and complete.

### Terraform Configuration Files
```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### Debug Output
Please provider a link to a GitHub Gist containing the complete debug output: https://www.terraform.io/docs/internals/debugging.html. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

### Expected Behavior
Lock should have been maintained

### Actual Behavior
Backend Init console output:
Initializing the backend...
 
Successfully configured the backend "consul"! Terraform will automatically
use this backend unless the backend configuration changes.
 
Terraform has been successfully initialized!

=======
Console output at end of provision:
aws_instance.TestNode01: Creation complete (ID: i-064182babeb197052)
 
Apply complete! Resources: 1 added, 0 changed, 1 destroyed.
 
The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
use the `terraform show` command.
 
State path:
 
Error releasing the state lock!
 
Error message: consul lock was lost
 
Terraform acquires a lock when accessing your state to prevent others
running Terraform to potentially modify the state at the same time. An
error occurred while releasing this lock. This could mean that the lock
did or did not release properly. If the lock didn't release properly,
Terraform may not be able to run future commands since it'll appear as if
the lock is held.
 
In this scenario, please call the "force-unlock" command to unlock the
state manually. This is a very dangerous operation since if it is done
erroneously it could result in two people modifying state at the same time.
Only call this command if you're certain that the unlock above failed and
that no one else is holding a lock.
 
1 error(s) occurred:
 
* Error releasing the state lock. Please see the longer error message above. 
======

### Steps to Reproduce
1. define consul backend
2. `terraform plan -out plan.file
3. terraform apply plan.file` (adding -lock-timeout=1600s does not change outcome)

### Important Factoids
Works fine with local or S3 backends.
Consul agent installed on machine running terraform and communicates with consul OK.

**Comment by curratore** [comment]
Same problem here with Terraform v0.9.5
Also tested with 0.9.3 and 0.9.4 versions with exactly the same behavior. 

The apply works perfectly and after a few secs the lock is released fine.

**Comment by jbardin** [comment]
Hi,

Sorry this is causing an issue, we definitely want consul support to be well supported in Terraform.

Is it possible for you to get debug logs from consul when this happens? You should be able to see when the session is created and the the session renew requests with the same ID up until the lock is released and the session destroyed. Unfortunately terraform isn't managing the consul session directly, so we don't log the actual ID if you need to extract it from a large log file. 

@curratore: When you say "after a few secs the lock is released fine", do you mean that you get the error shown above, but aren't blocked from running apply again?

**Comment by curratore** [comment]
Yes, exactly, I have the error but I can work without problems after the apply ends and the lock is released. It is quite weird behavior.

**Comment by jbardin** [comment]
@curratore, Thanks, that's actually the expected behavior. Consul locks are tied to the session, so once the session is lost the lock is freed by the server. 

Terraform is warning that the lock was lost, however the generic paragraph about force the unlock doesn't quite fit this situation. 

Without any logs to show otherwise, I can only assume there was some communication problem with the consul client+server that caused the session to be invalidated. Consul locks favor liveness, so if the terraform client loses the connection to the server the lock will be lost. I'm go to evaluate if this constraint can be relaxed for terraform, or possibly make the lock methodology configurable.

**Comment by curratore** [comment]
After upgrade to Terraform 0.9.6, I have the same behavior: 

```Apply complete! Resources: 73 added, 0 changed, 0 destroyed.

The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
use the `terraform show` command.

State path:

Error releasing the state lock!

Error message: consul lock was lost

Terraform acquires a lock when accessing your state to prevent others
running Terraform to potentially modify the state at the same time. An
error occurred while releasing this lock. This could mean that the lock
did or did not release properly. If the lock didn't release properly,
Terraform may not be able to run future commands since it'll appear as if
the lock is held.

In this scenario, please call the "force-unlock" command to unlock the
state manually. This is a very dangerous operation since if it is done
erroneously it could result in two people modifying state at the same time.
Only call this command if you're certain that the unlock above failed and
that no one else is holding a lock.

1 error(s) occurred:

* Error releasing the state lock. Please see the longer error message above.
```

Exactly the same happens, after I can work with this state without problems, but always shows the same error showed above.

**Comment by jbardin** [comment]
Hi @curratore,

This was closed yesterday, and the the patch will be included in the next release.

While this will allow Terraform to immediately try to reacquire the lock, and also verify the state hasn't been unexpectedly modified, it might be useful to figure out _why_ you're losing the locks in the first place. The most useful information is probably going to be from the consul server logs. 

I do plan to follow up and extend the terraform code to be able to get the session error and log it, but I'm not sure how helpful that will be compared to the consul logs.

**Comment by Boran** [comment]
I know this issues is closed but wanted to point out that v0.9.8 exhibits the same behaviour. I presume  the fix is inv0.9.9.

**Comment by jbardin** [comment]
@Boran,

The code to _better_ handle this is on v0.9.8. What we need to do is determine the root cause of _why_ you're losing the consul lock. Terraform can't prevent losing the lock on the consul server. 

What we really need are some consul logs from a system where the locks are lost during a terraform run to get a better idea of what's happening. 

I hope to improve the logging from the consul client in an upcoming release, which may help diagnose the problem, but chances are it's not going to provide as much useful information as what we get from the consul server itself.

**Comment by Boran** [comment]
I have access to the consul server (but am new to that too), so I can do specific logging or tracing if you like. What kind of logging do you want? In this thread or elsewhere?

**Comment by practicalint** [comment]
I have tried to use the consul backend again in Terraform v0.9.8 and still get the problem. I ran an aws instance creation with TRACE and attached the output here. I see the lock reacquire firing numerous times which may be a result of the change that was made, but the core problem is still there.

I do not see anything in the consul logs in /var/log/messages. Like Boran above, if you an give me some details from the Consul people on how to turn up/capture Consul logs, I could give you that.

I will also mention in case it is related that the machine running terraform and the instance being built both have the consul agent installed them as part of a standard build.

I will also note that this is running against and environment on the back-end, but the previous failures were not so I don't believe that would be related.

[MobaXterm_cfpmgtToolStation01_20170622_122618.docx](https://github.com/hashicorp/terraform/files/1095528/MobaXterm_cfpmgtToolStation01_20170622_122618.docx)

**Comment by jbardin** [comment]
Hi @practicalint,

Thanks for the log output. From Terraform's perspective, the lock was lost after exactly 60s, but the consul server still shows the lock as being active and doesn't allow Terraform to reacquire the lock.

The consul server may not have more information readily available without turning on debug logging. I'm going to run through some scenarios here and see if I can replicate this, or at least have some idea of what could be happening.

**Comment by gangofnuns** [comment]
FWIW, I'm having this same issue.    Not using environments, but using the remote consul server to store state.  

During a 'terraform plan', which takes quite awhile, I get the following.  
Running 'terraform force-unlock' fails with a crash and backtrace. 


[...]

Plan: 0 to add, 1 to change, 0 to destroy.

Error releasing the state lock!

Error message: consul lock was lost

Terraform acquires a lock when accessing your state to prevent others
running Terraform to potentially modify the state at the same time. An
error occurred while releasing this lock. This could mean that the lock
did or did not release properly. If the lock didn't release properly,
Terraform may not be able to run future commands since it'll appear as if
the lock is held.

**Comment by practicalint** [comment]
Hi @jbardin ,  the post yesterday with a confirmation of the issue reminded me that I was going to check in with you to see if you have been able to reproduce this. I have continued to try use the Consul backend while developing, but am moving off of it since it reports every run as failing and I'm needing to recognize true status now.  We would prefer Consul as the backend given our other uses and mix of HC products, so please continue to pursue this.
thanks

**Comment by jbardin** [comment]
Hi @practicalint,

Sorry this is still causing an issue for you. Now that we have most of the preliminary 0.10 work done I can get back to this issue. 

While I haven't been able to exactly reproduce your issue here, I do have some leads as to what it could be, and some ideas to try out to make the client more robust. 

This is definitely an issue I'd like to get fixed before the 0.10 release.

**Comment by sherabi** [comment]
Same issue in 0.9.11

**Comment by jantman** [comment]
Is it the case that the commit from July 17 still hasn't been released? I'm still having this on 0.10.2, but I can't find any reference to this issue in the changelog... so I can only assume it's still sitting in master unreleased?

**Comment by jbardin** [comment]
@jantman,

Tat commit has been included in the releases since 0.10.0. If you can replicate the problem, please feel free to open a new issue. Including the terraform and consul logs would be most helpful too. Thanks!
